### PR TITLE
Bugfix to handle broker resource state transitions to offline or dropped

### PR DESCRIPTION
**Pull request by antumbde**
Iterate over correct data structures to remove table while handling broker resource
state transition from online to offline/dropped state

### CODE REVIEW

**Comment by antumbde** [comment]
Original code was iterating over wrong data structures. It iterated over brokerRoutingTable which is keyed by tableName to find an entry in tableForInstance which is keyed by instance name. And then, it accessed the returned value without null check. That throws NPE.

There are no existing tests. I won't have time to start a new one for this complex piece. It should've been added when this code was redone.

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java [hunk 1]
```diff
@@ -519,16 +520,16 @@ public void markDataResourceOffline(String tableName) {
...
-        if (tablesForCurrentInstance.contains(tableName)) {
-          tablesForCurrentInstance.remove(tableName);
-        }
-
-        if (tablesForCurrentInstance.isEmpty()) {
-          _lastKnownInstanceConfigs.remove(instanceName);
+      Iterator<Map.Entry<String, Set<String>>> iterator = _tablesForInstance.entrySet().iterator();
+      while (iterator.hasNext()) {
+        Map.Entry<String, Set<String>> entry = iterator.next();
+        entry.getValue().remove(tableName);
```
Comment (Jackie-Jiang) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: Add a local variable to cache entry.getValue()?

**Comment by mcvsubbu** [comment]
Isn't there an issue in BrokerResourceOnlineOfflineStateModel as well? 
In the method onBecomeOnlineFromOffline() it seems we are using the broker resource name as table name?

**Comment by jfim** [comment]
As @mcvsubbu mentioned, the `onBecomeOfflineFromOnline` (and the dropped transitions) pass the resource name ("brokerResource") incorrectly, they should also pass the partition name (as is done correctly in `onBecomeOnlineFromOffline`).

**Comment by antumbde** [comment]
@jfim You should address the issue with onBecomeOfflineFromOnline in separate PR

**Pull request merged**